### PR TITLE
[DLS] Add "get_permissions" to BaseDataSource

### PR DESCRIPTION
**Pull request by timgrein**
## Closes https://github.com/elastic/enterprise-search-team/issues/4551

This PR adds the `get_permissions` method to the `BaseDataSource`, which should be implemented by connectors supporting DLS.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
~- [ ] Considered corresponding documentation changes~
~- [ ] Contributed any configuration settings changes to the configuration reference~

### CODE REVIEW

File: connectors/source.py [hunk 1]
```diff
@@ -467,6 +467,13 @@ async def close(self):
         """
         pass
 
+    async def get_permissions(self):
+        """Returns an asynchronous iterator on the permission documents present in the backend.
+
+        Each document is a dictionary containing permission data indexed into a corresponding permissions index.
+        """
+        raise NotImplementedError
```
Comment (wangch079) [inline connectors/source.py:475]: This implies every connector needs to implement a `get_permissions` method. Is it true?

**Review by wangch079 — COMMENTED** [review]
What is expected `get_permissions` implementation for a connector which doesn't support permissions

File: connectors/source.py (hunk 1 above)
Comment (timgrein) [reply connectors/source.py]: Returning an empty list now 👍 My idea was to guard the usage of this method with the corresponding feature flag so there's a "hard" error, when DLS is enabled via a manual index update, but not a strong opinion, so changing it.

Comment (wangch079) [reply connectors/source.py]: There's very similar scenario in incremental sync as well. Post it in [slack](https://elastic.slack.com/archives/C01795T48LQ/p1684913931191989) for wider discussion

Comment (timgrein) [reply connectors/source.py]: @seanstory Very valid points, either way I would guard these calls behind the feature flag, but as you've said people could update the flag manually and we would expect weird and unexpected behavior in the sense that anyone can look at a document. An error would be more safe IMO.

Thoughts @wangch079 @tarekziade ?

Comment (timgrein) [reply connectors/source.py:475]: Update: We had a discussion with @danajuratoni and from a product perspective it makes sense to hide the usage of `get_permissions` behind a feature flag and raise an error, when someone manually update the feature flag.

**Review by tarekziade — APPROVED** [review]

**Pull request merged**